### DISCUSSION TITLE
atom with set method in get

### DISCUSSION THREAD

**Discussion by elge70** [pmndrs/jotai#2283] [General] [answered]
In async atom get, if we want to store the data in another atom, there is no set method available in atom’s get method. Advantage of setting the data is that if data already exists we don’t have to fetch again in another component. atomWithCache does this. But storing in atom can isolate it from universal cache. One approach is to use store. Like store.set(<atom>,<value>) . In this method we cannot isolate to a component tree. Is there any other pattern I could use?

**Comment by dai-shi** [comment]
Doing `set` in the read function makes the atom data model "impure". So, it's intentionally prohibited.
That said, we have a new extension library for you needs.
Check out: https://github.com/jotaijs/jotai-effect

**Comment by elge70** [comment]
In my case, I am using atom family where depending on index, I update another atomfamily with same index. How can I incorporate index into Jotai-Effect.

**Reply by dai-shi** [reply]
I think you can create another atom family combining them. Just like you can do with two atoms.

```js
const atom1 = atom(1)
const atom2 = atom(2)
const atom3 = atom(
  (get) => ...,
  (get, set, arg) => ...
)
```

**Comment by elge70** [comment]
I am not sure how a third atom will solve the issue. When 3rd atom trying to get data, the fetch function may fetch the data. To illustrate my point  sample code is attached 

```
export const myStore = createStore();

const getData = async (props: string) => {
  const url = "http://localhost:5000/users/" + props;
  return (await fetch(url)).json();
};

export interface GraphParam {
  id: string;
  url: string;
}

export interface GraphData {
  data: string | undefined;
}

// Map is from immutable.js
// URL is set in this atomFamily
export const urlList = atom<Map<string, GraphParam>>(Map<string, GraphParam>());
// Data for URL is stored here
export const dataList = atom<Map<string, GraphData>>(Map<string, GraphData>());

export const graphParamAF = atomFamily((key: string) =>
  atom(
    (get) => {
      const tval = get(urlList);
      if (tval.has(key)) {
        return tval.get(key);
      } else {
        return { id: key, url: "Undefined" };
      }
    },
    (get, set, arg: GraphParam) => {
      const pval = get(urlList);
      const param = pval.get(key);
      //  isEqual from "react-fast-compare";
      if (isEqual(param, arg)) {
        console.log("Not Setting....Equal Params....", arg);
      } else {
        console.log("Setting New Value.......", arg);
        set(urlList, pval.set(key, arg));
        const dval = get(dataList);
        // New URL set or to re-fetch the data
        // delete the data for URL to refresh the data
        set(dataList, dval.delete(key));
      }
    },
  ),
);

export const graphDataAF = atomFamily((key: string) =>
  atom(async (get) => {
    const turl = get(urlList);
    const url = turl.get(key);
    const dval = get(dataList);
    if (dval.has(key)) {
      // cached and don't refresh the data
      return dval.get(key);
    }
    if (url?.url != "Undefined") {
      // refresh the data for new URL or refresh for same URL
      const udata = await getData(url?.url as string);
      const data = get(dataList);
      // this is what I want to avoid. This works fine
      myStore.set(dataList, data.set(key, udata));
      return udata;
    } else {
      return { data: undefined };
    }
  }),
);
```

**Reply by dai-shi** [reply]
Ah, so you use atomWithRefresh pattern for Suspense.
dataList is a simple cache for getData, isn't it? I wonder moving it out of atom model makes it easier.

**Reply by elge70** [reply]
I am using the dataList as cache. But the url parameters are set in one component and the data is fetched in another component depending the URL change. This data is used to render the chart.  If I move the dataList from atom, how will  the component will refresh based on new url  or data update ? I am using suspense and error boundary for the fetch and display component.

**Reply by dai-shi** [reply]
Isn't new url in atom?
If I understand correctly, your problem isn't tied to atomFamily.
Would you like to create a smaller example to isolate your problem at minimum?

**Reply by elge70** [reply]
States in my application is in atomFamily. This is a dashboard application rendered dynamically based on Json data. Therefore the URL parameters are in atomFamily and the data is in another atomFamily. Each chart gets its url using its "id" and gets the data and store it another atomFamily with same id. The number of charts and its filters are dynamic based on Json data.

Due to dynamic nature, I found atomFamily a good solution.  atomEffect is a good solution if I can pass the "id" along with get & set method similar to update.


```
const loggingEffect = atomEffect((get, set) => {
  // runs on mount or whenever someAtom changes
  const value = get(someAtom)  // Here get the appropriate atom from another family
  loggingService.setValue(value)
})
```
Whether you could extend to pass index parameter to access atomFamily ?

**Reply by dai-shi** [reply]
Using atomFamily in your application is just fine. I'm not very confident if I understand your problem correctly, so if it's not atomFamily specific problem, it would be easier to understand the problem and discuss a solution without atomFamily.

**Reply by elge70** [reply]
Following is the simplified version. Here I have a one data component sets the URL parameters and two component displays the values. One component displays in Chart and other in Table format. Both component uses the same atom to get data. In my application all are atomFamilies since I have many such components generated at runtime. My requirement is 

Once data is fetched in graphData atom, it should not fetch again unless URL parameters change. Same data is used in 2 components.

1. atomEffect is a good solution. How to extend for a family of atoms
2. jotai-cache is also a good solution. How to extend for a family of atoms. Also if you could add a _shouldRemove_ function example , it would really help.
3. If you could add a hint for patterns that can be extended for atomFamily, it is of great help.

```

import { Button } from "@mantine/core";
import { atom, createStore, useAtomValue, useSetAtom } from "jotai";
import superjson from "superjson";

export const myStore = createStore();

const getData = async (props: string) => {
  const url = "http://localhost:5000/users/" + props;
  return (await fetch(url)).json();
};

export interface GraphParam {
  id: string;
  url: string;
}

export interface GraphData {
  data: string | undefined;
}

export const graphParam = atom<GraphParam>({ id: "", url: "" });

export const graphData = atom(async (get) => {
  const param = get(graphParam);
  const udata = await getData(param?.url as string);
  return udata;
});

export const DataInputComponent = () => {
  const setParam = useSetAtom(graphParam);

  const buttonClicked = () => {
    setParam({ id: "1", url: Math.floor(Math.random() * 1000).toString() });
  };
  return <Button onClick={buttonClicked}>GetData </Button>;
};

export const ChartComponent = () => {
  const data = useAtomValue(graphData);
  return <div>{superjson.stringify(data)} </div>;
};

export const TableComponent = () => {
  const data = useAtomValue(graphData);
  return <div>{superjson.stringify(data)} </div>;
};

```

**Reply by dai-shi** [reply]
Thanks for the code snippet.

> Once data is fetched in graphData atom, it should not fetch again unless URL parameters change.

Do you mean you want to avoid refetching if `param.url` is the same, but `param.id` is different?
Otherwise, it seems pretty stable (= not refetch even with two components.)
Or, maybe you are talking about suspense waterfall.

Either case, we have a solution. So, I'll wait for clarifying the problem.

**Comment by elge70** [comment]
For a given component (in this case Table and Chart component which displays the details of a device) the id remains the same. Since there are many such display components, each will have its own ids. When the url changes depending on the filter parameters, id will remain the same (Different components will have different URLs and IDs). Once the url changes, we need to fetch again one for each component ( It should not fetch twice for Chart and Table component since both are using same atom).  

Another scenario, I want to handle is refreshing the Chart & Table component when URL remains the same. In this case if the data is empty, it should fetch the data. 

Each component (Table and Chart) is inside suspense and error boundary to catch errors.

**Reply by dai-shi** [reply]
So, I think it's the second problem.
Can you make sure if the problem only happens with async atom?

For example, changing this doesn't cause the problem?
```js
export const graphData = atom((get) => {
  const param = get(graphParam);
  const udata = "dummy_data_" + Math.random();
  return udata;
});
```

In such a case, it's a known problem, sometimes called suspense waterfalls.
The problem is not specific to Jotai, but we have a workaround.
See this thread: https://twitter.com/dai_shi/status/1509533380786155525

**Comment by elge70** [comment]
Thanks. I will check this scenario. How can I handle refresh scenario where URL is not changing, and I want to clear data and fetch the data.

**Reply by dai-shi** [reply]
https://jotai.org/docs/recipes/atom-with-refresh should do.

**Comment by elge70** [comment]
The solution may work for single atom (not yet tried) but not for atomFamily
```

export const graphDataAF = atomFamily((key: string) =>
  atom(async (get) => {
    console.log("Getting a GET for request ...... Key => ", key);
    const turl = get(urlList);
    const url = turl.get(key);
// I tried  const url = get(urlList).get(key); and no change in refreshing...
    if (url?.url != "Undefined") {
      console.log("Getting value for url --> ", url?.url);
      const udata = await getData(url?.url as string);
      return { data: udata };
    } else {
      return { data: undefined };
    }
  }),
);

```
When I change an URL in the url atomFamily, entire lists of components are refreshing even though the URLs for other components have not changed. This is due to this code 

```
    const turl = get(urlList);
    const url = turl.get(key);
```

The list has changed due to single URL change but not the other URLs. Any solution ? Also whether you can combine atom-with-refresh with atomWithPending  as I need refresh + Async with Suspence

**Reply by dai-shi** [reply]
Using `Map` correctly requires updating maps immutably. `new Map(prev).set(key, val)`

It seems like you have several isolated issues. I can help each, but please try to combine those ideas on your end.

**Reply by elge70** [reply]
I am using immutable.js Map which gives a new Map every time.

**Reply by dai-shi** [reply]
Oh I see! Then, I'm not sure why the change doesn't trigger updates.

**Reply by elge70** [reply]
It is triggering updates for all components even though the URL has changed only for one component. Basically, List has changed, and all components are fetching data.

**Reply by dai-shi** [reply]
If it's because of suspense waterfalls, jotai-scope is a workaround. Otherwise, you need to dig into further.
To confirm if it's suspense waterfalls, changing the async atom to the sync atom and the problem should be gone.

**Reply by elge70** [reply]
I made the fetch data synchronous 
const getData = (props: string) => {
  const url = "http://localhost:5000/users/" + props;
  const request = new XMLHttpRequest();
  request.open("GET", url, false);
  request.send(null);

  if (request.status === 200) {
    return request.responseText;
  }
  return null;
};

Still it is refreshing all components.

**Reply by dai-shi** [reply]
I wonder if you can create a very very small reproduction without async atoms without atomFamily.

**Reply by elge70** [reply, marked as answer]
I figured out the issue.  I should have used atomFamily of URLs in the async atomFamily. It is working great now. It did not create any problem in Suspense & Error Boundary. Thanks for all the help.

Also is it possible to get the store inside the atom methods. getDefaultStore get the default provider. This will not give store inside the provider. 

atomEffect is great utility. Is it possible to extend so that it could be used with atomFamily ?

**Reply by dai-shi** [reply]
atom should be pure functions. access to the store is impure.
Not sure about atomEffect and atomFamily. I would generally avoid atomFamily, which is outside of "atom" model.

**Reply by elge70** [reply]
In atoms-in-atom can we dynamically add atoms which can replicate atomFamily ?

in https://jotai.org/docs/guides/atoms-in-atom 
```
const pricesAtom = atom({
  apple: atom(15),
  orange: atom(12),
  pineapple: atom(25),
})
```
Could we add new atoms here dynamically ?

**Reply by dai-shi** [reply]
Yes:
```js
  const setPrices = useSetAtom(pricesAtom);
  const handleClick = () => {
    setPrices((prev) => ({ ...prev, grape: atom(35) }));
  };
```

**Reply by elge70** [reply]
If I want to refresh the data in atomFamily whether I could extend atomWithRefresh in atomFamily 
```
export const graphDataAF = atomFamily((key: string) =>
  atom(async (get) => {
    console.log("Getting a GET for request ...... Key => ", key);
    const url = get(graphParamAF(key));
    if (url?.url != "Undefined") {
      console.log("Getting value for url --> ", url?.url);
      const udata = await getData(url?.url as string);
      return udata;
    } else {
      return { data: undefined };
    }
  }),
);

```

**Reply by dai-shi** [reply]
Do you mean something like this?
```js
const refreshAF = atomFamily((key) => atom(0));
const dataAF = atomFamily((key) => atom(async (get) => {
  get(refreshAF(key));
  // ...
}));
```

**Reply by elge70** [reply]
```

export const graphDataAF = atomFamily((key: string) =>
  atom(
    async (get) => {
      console.log("Getting a GET for request ...... Key => ", key);
      const url = get(graphParamAF(key));
      if (url?.url != "Undefined") {
        console.log("Getting value for url --> ", url?.url);
        const udata = await getData(url?.url as string);
        return udata;
      } else {
        return { data: undefined };
      }
    },
    async (get, set) => {
      console.log("Getting a GET for request ...... Key => ", key);
      const url = get(graphParamAF(key));
      if (url?.url != "Undefined") {
        console.log("Getting value for url --> ", url?.url);
        const udata = await getData(url?.url as string);
		  // set the data for the atom here !
      } else {
          //set data here  { data: undefined };
      }
    },
  ),
);

const[data,refreshData] = useAtom(graphDataAF(key))


```
Is there any way to set the value of atom within the atom ?

**Reply by dai-shi** [reply]
If I'm not mistaken, it would be something like this?
```js
const dataAF = atomFamily((key) => {
  const baseAtom = atomWithDefault(/* ... */);
  return atom(
    (get) => get(baseAtom),
    async (get, set) => {
      // ...
    }
  }
);
```

**Reply by elge70** [reply]
Working perfectly! I must use useTransition to handle suspense. Thanks a lot!!!

**Reply by elge70** [reply]
whether async write in atom set function can be used with https://github.com/pmndrs/suspend-react ? Are there any recipe for integration

**Reply by dai-shi** [reply]
suspend-react will be superseded by `use`. I have a similar library https://github.com/dai-shi/react-suspense-fetch which is marked as deprecated.

Related discussion about using `use` directly: #2293
